From ea3e318b89c42d2eff2ce0b9d78dc2371fbb6a67 Mon Sep 17 00:00:00 2001
From: Hendrik Leppkes <h.leppkes@gmail.com>
Date: Tue, 21 Jul 2026 18:27:56 +0200
Subject: [PATCH] hdmv: fix INSN_BC instruction

The condition was backwards, checking if any bits outside of the desired
are set, rather then the desired bits.

This still differs from the specification which just indicates a simple
AND check, however the spec does also not indicate if the result should
be checked to be non-zero, or matching the input.

This version is likely what was originally intended, however all discs
that were found to use the BC instruction only test a single bit, making
the difference between a full or partial match not testable.
---
 src/libbluray/hdmv/hdmv_vm.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/libbluray/hdmv/hdmv_vm.c b/src/libbluray/hdmv/hdmv_vm.c
index c2c60be9..e1b5a817 100644
--- a/src/libbluray/hdmv/hdmv_vm.c
+++ b/src/libbluray/hdmv/hdmv_vm.c
@@ -1134,7 +1134,7 @@ static int _hdmv_step(HDMV_VM *p)
                 BD_DEBUG(DBG_HDMV|DBG_CRIT, "missing operand in BRANCH/JUMP opcode 0x%08x\n", *(uint32_t*)insn);
             }
             switch (insn->cmp_opt) {
-                case INSN_BC: p->pc += !!(dst & ~src); break;
+                case INSN_BC: p->pc += !!(src & ~dst); break;
                 case INSN_EQ: p->pc += !(dst == src); break;
                 case INSN_NE: p->pc += !(dst != src); break;
                 case INSN_GE: p->pc += !(dst >= src); break;
